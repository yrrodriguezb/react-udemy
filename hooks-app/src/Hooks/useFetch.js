import { useEffect, useRef, useState } from "react"


export const useFetch = (url) => {
  const isMounted = useRef(true)

  useEffect(() => {
    return () => {
      isMounted.current = false;
    }
  }, []) // Dependencia vacia, que solo lo haga la primera vez que el componente se carga

  const [state, setState] = useState({
    data: null, 
    loading: true, 
    error: null
  })

  useEffect(() => {
    setState({
      loading: true,
      data: null,
      error: null     
    })

    fetch(url)
      .then(res => res.json())
      .then(data => {
        // evita que ocuarra un error cuando el componente se ha desmontado y no se ha solucionado la promesa
        // para setear el nuevo estado. setTimeout para pruebas y replica del error
        // setTimeout(() => {
          if (isMounted.current) {
            setState({
              loading: false,
              error: null,
              data: data
            })
          } else {
            console.info(`[${Date.now().toString()}]: No se puede establacer el nuevo estado del useFetch porque el componente fue desmontado`)
          }
        // }, 4000);
      }) 
  }, [ url ])

  return state;
}
