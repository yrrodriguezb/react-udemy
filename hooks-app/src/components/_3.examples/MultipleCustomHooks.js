import React from 'react'
import { useCounter } from '../../Hooks/useCounter';
import { useFetch } from '../../Hooks/useFetch'

export const MultipleCustomHooks = () => {
  const { counter, increment } =  useCounter(1);

  const state =  useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`)

  const { loading, data } = state;
  const { author, quote } = !!data && data[0];

 

  return (
    <div className='m-5'>
      <h1>BreakingBad Quotes</h1>

      {
        loading 
        ?
          (
            <div className='alert alert-info text-center'>
              Loading...
            </div>
          )
        :
          (
            <blockquote className='blockquote text-right'>
              <p className='mb-1'>{ quote }</p>
              <footer className='blockquote-footer'>
                { author }
              </footer>
            </blockquote>
          )
      }

      <button 
        className='btn btn-primary'
        onClick={ increment }  
      >
        Siguiente quote
      </button>
    </div>
  )
}
