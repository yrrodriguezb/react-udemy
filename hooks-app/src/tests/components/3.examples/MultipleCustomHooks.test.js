import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/_3.examples/MultipleCustomHooks'
import { useFetch } from '../../../Hooks/useFetch';
import { useCounter } from '../../../Hooks/useCounter';

// Usa una implementacion del useFetch para simular los datos
jest.mock('../../../Hooks/useCounter')
jest.mock('../../../Hooks/useFetch')


describe('Pruebas en <MultipleCustomHooks />', () => { 

  useCounter.mockReturnValue({
    counter: 10,
    increment: () => {}
  });

  test('Debe mostrarse correctamente', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    })

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Debe mostrar la informaciÃ³n', () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: 'Yeison',
          quote: 'Hola Mundo'
        }
      ],
      loading: false,
      error: null
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    // console.log(wrapper.html());
    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('p.mb-1').text().trim()).toBe('Hola Mundo');
    expect(wrapper.find('footer').text().trim()).toBe('Yeison');
  });
});