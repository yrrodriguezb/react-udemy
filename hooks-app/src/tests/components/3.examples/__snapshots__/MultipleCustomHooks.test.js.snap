// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> Debe mostrarse correctamente 1`] = `
<div
  className="m-5"
>
  <h1>
    BreakingBad Quotes
  </h1>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
